doc/sphinx/user-extensions/syntax-extensions.rst

```diff
@@ -1701,131 +1592,183 @@ String notations
+
+   For example, the following command defines a notation with a single parameter `x`.
+
+   .. coqtop:: in
 
-   .. cmd:: Print Grammar tactic
+      Tactic Notation "destruct_with_eqn" constr(x) := destruct x eqn:?.
 
-      To know the parsing precedences of the existing tacticals, use the command
-      ``Print Grammar tactic``.
+   For a complex example, examine the 16 `Tactic Notation "setoid_replace"`\s
+   defined in :file:`$COQLIB/theories/Classes/SetoidTactics.v`, which are designed
+   to accept any subset of 4 optional parameters.
 
-   Each type of tactic argument has a specific semantic regarding how it
-   is parsed and how it is interpreted. The semantic is described in the
-   following table. The last command gives examples of tactics which use
-   the corresponding kind of argument.
+   The nonterminals that can specified in the tactic notation are:
+
+   .. todo PR At the moment I splice uconstr.  Should I bring it back?
+      An untyped term is one that doesn't get typechecked??
+
+   .. todo 'open_constr' appears to be another possible value based on the
+      the message from "Tactic Notation open_constr := idtac".
+      Also (at least) "ref", "string", "preident", "int" and "ssrpatternarg".
+      (from reading .v files).
+      Looks like any string passed to "make0" in the code is valid.  But do
+      we want to support all these?
 
    .. list-table::
       :header-rows: 1
 
-      * - Tactic argument type
-        - parsed as
-        - interpreted as
+      * -  Specified :token:`ident`
+        - Parsed as
+        - Interpreted as
         - as in tactic
 
       * - ``ident``
-        - identifier
+        - :token:`ident`
         - a user-given name
-        - intro
+        - :tacn:`intro`
 
       * - ``simple_intropattern``
-        - simple_intropattern
+        - :token:`simple_intropattern`
         - an introduction pattern
-        - assert as
+        - :tacn:`assert` `as`
 
       * - ``hyp``
-        - identifier
+        - :token:`ident`
         - a hypothesis defined in context
-        - clear
+        - :tacn:`clear`
 
       * - ``reference``
-        - qualified identifier
+        - :token:`qualid`
         - a global reference of term
-        - unfold
+        - :tacn:`unfold`
 
       * - ``constr``
-        - term
+        - :token:`term`
         - a term
-        - exact
+        - :tacn:`exact`
 
       * - ``uconstr``
-        - term
+        - :token:`term`
         - an untyped term
-        - refine
+        - :tacn:`refine`
 
       * - ``integer``
-        - integer
+        - :token:`int`
         - an integer
         -
 
       * - ``int_or_var``
-        - identifier or integer
+        - :token:`int_or_var`
         - an integer
-        - do
+        - :tacn:`do`
 
       * - ``tactic``
-        - tactic at level 5
+        - :token:`ltac_expr`
         - a tactic
         -
 
       * - ``tacticn``
-        - tactic at level n
+        - :token:`ltac_expr` at level n
```

Comment:
Is it valid to use levels other than 0..5?  I think it might not be, and somehow it should be specified that these are the valid/common levels.

Reply:
Good point.  `tactic6` is not accepted.  I changed this entry to look like this:

![image](https://user-images.githubusercontent.com/1253341/80320674-acc6a880-87cc-11ea-8021-1c51bc71b4da.png)

Reply:
I don't see the change here in the PR.  And I think the second column should still be :token:`ltac_expr` at level n

Reply:
You will when I push :-).  Usually I push updates to a PR no more than once a day.  Still waiting for any further comments from you.  I'll plan to squash and push around 9 PM PDT this evening.

> I think the second column should still be :token:`ltac_expr` at level n.

Well, even the old, not updated chapter has ltac_expr0..3.  While we'll need to mention levels in a few places in the doc (to be added), we're presenting the grammar without making level a central concept.  At that time we'll put `level` in the glossary with a hyperlink.  (I expect the glossary will begin to appear in the doc sometime this week--with only a few definitions to begin with.)

If you still feel strongly about the second column I will change it back.

Reply:
Ah, if `ltac_exprn` is used in other places in the text, then using it here seems fine.

Reply:
@jfehrle Regarding the glossary, and the presentation of the grammar, have you noticed #12172?